File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/infra/adapter/ApiAdapter.java
/*
 * Copyright © 2015 The Gravitee team (http://gravitee.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.gravitee.apim.infra.adapter;

import io.gravitee.apim.core.api.model.Api;
import io.gravitee.apim.core.api.model.crd.ApiCRDSpec;
import io.gravitee.apim.core.membership.model.PrimaryOwnerEntity;
import io.gravitee.definition.model.DefinitionVersion;
import io.gravitee.definition.model.federation.FederatedAgent;
import io.gravitee.definition.model.federation.FederatedApi;
import io.gravitee.definition.model.v4.ApiType;
import io.gravitee.definition.model.v4.nativeapi.NativeApi;
import io.gravitee.rest.api.model.context.OriginContext;
import io.gravitee.rest.api.model.federation.FederatedApiAgentEntity;
import io.gravitee.rest.api.model.federation.FederatedApiEntity;
import io.gravitee.rest.api.model.v4.api.ApiEntity;
import io.gravitee.rest.api.model.v4.api.GenericApiEntity;
import io.gravitee.rest.api.model.v4.api.NewApiEntity;
import io.gravitee.rest.api.model.v4.api.UpdateApiEntity;
import io.gravitee.rest.api.model.v4.nativeapi.NativeApiEntity;
import java.io.IOException;
import java.util.Collection;
import java.util.Map;
import java.util.stream.Stream;
import org.mapstruct.DecoratedWith;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.MappingConstants;
import org.mapstruct.ValueMapping;
import org.mapstruct.factory.Mappers;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

@Mapper(uses = { PlanAdapter.class })
@DecoratedWith(ApiAdapterDecorator.class)
public interface ApiAdapter {
    Logger log = LoggerFactory.getLogger(ApiAdapter.class);
    ApiAdapter INSTANCE = Mappers.getMapper(ApiAdapter.class);

    @Mapping(target = "apiDefinitionHttpV4", expression = "java(deserializeApiDefinitionV4(source))")
    @Mapping(target = "apiDefinition", expression = "java(deserializeApiDefinitionV2(source))")
    @Mapping(target = "federatedApiDefinition", expression = "java(deserializeFederatedApiDefinition(source))")
    @Mapping(target = "apiDefinitionNativeV4", expression = "java(deserializeNativeApiDefinition(source))")
    Api toCoreModel(io.gravitee.repository.management.model.Api source);

    Stream<Api> toCoreModelStream(Stream<io.gravitee.repository.management.model.Api> source);

    @Mapping(target = "definition", expression = "java(serializeApiDefinition(source))")
    io.gravitee.repository.management.model.Api toRepository(Api source);

    Stream<io.gravitee.repository.management.model.Api> toRepositoryStream(Stream<Api> source);

    @Mapping(target = "apiVersion", source = "version")
    @Mapping(target = "tags", source = "apiDefinitionHttpV4.tags")
    @Mapping(target = "listeners", source = "apiDefinitionHttpV4.listeners")
    @Mapping(target = "endpointGroups", source = "apiDefinitionHttpV4.endpointGroups")
    @Mapping(target = "analytics", source = "apiDefinitionHttpV4.analytics")
    @Mapping(target = "flowExecution", source = "apiDefinitionHttpV4.flowExecution")
    @Mapping(target = "flows", source = "apiDefinitionHttpV4.flows")
    @Mapping(target = "failover", source = "apiDefinitionHttpV4.failover")
    NewApiEntity toNewApiEntity(Api source);

    @ValueMapping(source = MappingConstants.ANY_REMAINING, target = MappingConstants.NULL)
    @Mapping(source = "version", target = "apiVersion")
    @Mapping(target = "disableMembershipNotifications", expression = "java(!spec.isNotifyMembers())")
    @Mapping(target = "listeners", expression = "java((List<Listener>) spec.getListeners())")
    @Mapping(target = "endpointGroups", expression = "java((List<EndpointGroup>) spec.getEndpointGroups())")
    @Mapping(target = "flows", expression = "java((List<Flow>) spec.getFlows())")
    UpdateApiEntity toUpdateApiEntity(ApiCRDSpec spec);

    @ValueMapping(source = MappingConstants.ANY_REMAINING, target = MappingConstants.NULL)
    @Mapping(target = "id", source = "api.id")
    @Mapping(target = "name", source = "api.name")
    @Mapping(target = "definitionVersion", source = "api.definitionVersion")
    @Mapping(target = "type", source = "api.type")
    @Mapping(target = "tags", source = "api.tags")
    @Mapping(target = "apiVersion", source = "api.version")
    @Mapping(target = "lifecycleState", source = "api.apiLifecycleState")
    UpdateApiEntity toUpdateApiEntity(Api api, io.gravitee.definition.model.v4.Api apiDefinitionV4);

    @ValueMapping(source = MappingConstants.ANY_REMAINING, target = MappingConstants.NULL)
    @Mapping(source = "version", target = "apiVersion")
    @Mapping(target = "metadata", ignore = true)
    @Mapping(target = "listeners", expression = "java((List<Listener>) api.getListeners())")
    ApiEntity toApiEntity(ApiCRDSpec api);

    @ValueMapping(source = MappingConstants.ANY_REMAINING, target = MappingConstants.NULL)
    @Mapping(source = "version", target = "apiVersion")
    @Mapping(target = "metadata", ignore = true)
    @Mapping(target = "listeners", expression = "java((List<Listener>) api.getApiListeners())")
    ApiEntity toApiEntity(Api api);

    @ValueMapping(source = MappingConstants.ANY_REMAINING, target = MappingConstants.NULL)
    @Mapping(source = "version", target = "apiVersion")
    @Mapping(target = "metadata", ignore = true)
    @Mapping(target = "tags", source = "apiDefinitionNativeV4.tags")
    @Mapping(target = "listeners", source = "apiDefinitionNativeV4.listeners")
    @Mapping(target = "endpointGroups", source = "apiDefinitionNativeV4.endpointGroups")
    @Mapping(target = "flows", source = "apiDefinitionNativeV4.flows")
    @Mapping(target = "resources", source = "apiDefinitionNativeV4.resources")
    @Mapping(target = "services", source = "apiDefinitionNativeV4.services")
    @Mapping(target = "properties", source = "apiDefinitionNativeV4.properties")
    @Mapping(target = "state", source = "lifecycleState")
    @Mapping(target = "lifecycleState", source = "apiLifecycleState")
    NativeApiEntity toNativeApiEntity(Api api);

    @ValueMapping(source = MappingConstants.ANY_REMAINING, target = MappingConstants.NULL)
    @Mapping(source = "source.version", target = "apiVersion")
    @Mapping(source = "source.id", target = "id")
    @Mapping(target = "primaryOwner", source = "primaryOwnerEntity")
    @Mapping(target = "referenceId", source = "source.environmentId")
    @Mapping(target = "referenceType", constant = "ENVIRONMENT")
    @Mapping(source = "source.apiLifecycleState", target = "lifecycleState")
    FederatedApiEntity toFederatedApiEntity(io.gravitee.repository.management.model.Api source, PrimaryOwnerEntity primaryOwnerEntity);

    @ValueMapping(source = MappingConstants.ANY_REMAINING, target = MappingConstants.NULL)
    @Mapping(target = "apiVersion", source = "source.version")
    @Mapping(target = "id", source = "source.id")
    @Mapping(target = "primaryOwner", source = "primaryOwnerEntity")
    @Mapping(target = "referenceId", source = "source.environmentId")
    @Mapping(target = "referenceType", constant = "ENVIRONMENT")
    @Mapping(target = "lifecycleState", source = "source.apiLifecycleState")
    @Mapping(target = "provider", source = "agent.provider")
    @Mapping(target = "defaultInputModes", source = "agent.defaultInputModes")
    @Mapping(target = "defaultOutputModes", source = "agent.defaultOutputModes")
    @Mapping(target = "capabilities", expression = "java(capabilities(agent))")
    @Mapping(target = "skills", source = "agent.skills")
    @Mapping(target = "securitySchemes", source = "agent.securitySchemes")
    @Mapping(target = "security", source = "agent.security")
    @Mapping(target = "name", source = "source.name")
    @Mapping(target = "definitionVersion", source = "source.definitionVersion")
    @Mapping(target = "description", source = "source.description")
    @Mapping(target = "originContext", source = "originContext")
    FederatedApiAgentEntity toFederatedAgentEntity(
        io.gravitee.repository.management.model.Api source,
        FederatedAgent agent,
        PrimaryOwnerEntity primaryOwnerEntity,
        OriginContext.Integration originContext
    );

    default Collection<String> capabilities(FederatedAgent agent) {
        return agent == null || agent.getCapabilities() == null
            ? null
            : agent.getCapabilities().entrySet().stream().filter(Map.Entry::getValue).map(Map.Entry::getKey).toList();
    }

    @Mapping(source = "state", target = "lifecycleState")
    @Mapping(source = "lifecycleState", target = "apiLifecycleState")
    @ValueMapping(source = MappingConstants.ANY_REMAINING, target = MappingConstants.NULL)
    Api fromApiEntity(GenericApiEntity apiEntity);

    default <T> T deserialize(io.gravitee.repository.management.model.Api api, Class<T> clazz) {
        if (api.getDefinition() == null) {
            log.debug(
                "The API definition is null for api id {}. This can happen when filtering the definition using ApiFieldFilter",
                api.getId()
            );
            return null;
        }
        try {
            return GraviteeJacksonMapper.getInstance().readValue(api.getDefinition(), clazz);
        } catch (IOException ioe) {
            log.error("An error occurred while deserializing V4 API definition for api id {}", api.getId(), ioe);
            return null;
        }
    }

    default io.gravitee.definition.model.v4.Api deserializeApiDefinitionV4(io.gravitee.repository.management.model.Api api) {
        return api.getDefinitionVersion() == DefinitionVersion.V4 && api.getType() != ApiType.NATIVE
            ? deserialize(api, io.gravitee.definition.model.v4.Api.class)
            : null;
    }

    default io.gravitee.definition.model.Api deserializeApiDefinitionV2(io.gravitee.repository.management.model.Api api) {
        return api.getDefinitionVersion() == DefinitionVersion.V2 || api.getDefinitionVersion() == null
            ? deserialize(api, io.gravitee.definition.model.Api.class)
            : null;
    }

    default io.gravitee.definition.model.federation.FederatedApi deserializeFederatedApiDefinition(
        io.gravitee.repository.management.model.Api api
    ) {
        return api.getDefinitionVersion() == DefinitionVersion.FEDERATED ? deserialize(api, FederatedApi.class) : null;
    }

    default NativeApi deserializeNativeApiDefinition(io.gravitee.repository.management.model.Api api) {
        return api.getDefinitionVersion() == DefinitionVersion.V4 && api.getType() == ApiType.NATIVE
            ? deserialize(api, NativeApi.class)
            : null;
    }

    default String serializeApiDefinition(Api api) {
        return switch (api.getDefinitionVersion()) {
            case V1, V2 -> serialize(api.getApiDefinition(), "V2 API");
            case V4 -> switch (api.getType()) {
                case NATIVE -> serialize(api.getApiDefinitionNativeV4(), "V4 Native API");
                case PROXY, MESSAGE -> serialize(api.getApiDefinitionHttpV4(), "V4 API");
            };
            case FEDERATED -> serialize(api.getFederatedApiDefinition(), "Federated API");
            case FEDERATED_AGENT -> serialize(api.getFederatedAgent(), "Federated Agent");
        };
    }

    default <T> String serialize(T value, String name) {
        try {
            return GraviteeJacksonMapper.getInstance().writeValueAsString(value);
        } catch (IOException ioe) {
            throw new RuntimeException("Unexpected error while serializing " + name + " definition", ioe);
        }
    }
}

File: gravitee-apim-console-webui/src/shared/components/zee/adapters/api-adapter.ts
/*
 * Copyright (C) 2015 The Gravitee team (http://gravitee.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ZeeResourceAdapter } from '../zee.model';

interface ApiFlowStep {
  name: string;
  policy: string;
  enabled: boolean;
  description?: string;
  condition?: string;
  configuration: Record<string, unknown>;
}

interface ApiFlow {
  name: string;
  enabled: boolean;
  selectors: unknown[];
  request: ApiFlowStep[];
  response: ApiFlowStep[];
  subscribe: ApiFlowStep[];
  publish: ApiFlowStep[];
  tags: string[];
}

interface ApiEntrypoint {
  type: string;
  configuration: Record<string, unknown>;
  dlq: { endpoint: string } | null;
}

interface ApiListener {
  type: string;
  entrypoints: ApiEntrypoint[];
  paths: unknown[];
  servers: unknown[];
  cors?: unknown | null;
}

interface ApiEndpoint {
  name: string;
  type: string;
  weight: number;
  inheritConfiguration: boolean;
  secondary: boolean;
  configuration: Record<string, unknown>;
  services: Record<string, unknown>;
}

interface ApiEndpointGroup {
  name: string;
  type: string;
  loadBalancer: { type: string };
  endpoints: ApiEndpoint[];
  sharedConfiguration: Record<string, unknown>;
  services: Record<string, unknown>;
}

interface ApiProperty {
  key: string;
  value: string;
}

interface ApiPayload {
  name: string;
  apiVersion: string;
  description: string;
  type: string;
  listeners: ApiListener[];
  endpointGroups: ApiEndpointGroup[];
  flows: ApiFlow[];
  properties: ApiProperty[];
  tags: string[];
}

export const API_ADAPTER: ZeeResourceAdapter<ApiPayload> = {
  previewLabel: 'Generated API',
  transform: (generated: unknown): ApiPayload => {
    const g = generated as Record<string, unknown>;
    return {
      name: asString(g.name, ''),
      apiVersion: asString(g.apiVersion ?? g.version, '1.0.0'),
      description: asString(g.description, ''),
      type: asString(g.type, 'PROXY'),
      listeners: asArray(g.listeners).map(mapListener),
      endpointGroups: asArray(g.endpointGroups).map(mapEndpointGroup),
      flows: asArray(g.flows).map(mapFlow),
      properties: asArray(g.properties).map(mapProperty),
      tags: asStringArray(g.tags),
    };
  },
};

function mapListener(raw: unknown): ApiListener {
  const l = raw as Record<string, unknown>;
  const type = asString(l.type, '');
  return {
    type,
    entrypoints: asArray(l.entrypoints).map(mapEntrypoint),
    paths: asArray(l.paths),
    servers: asArray(l.servers),
    ...(type === 'HTTP' ? { cors: l.cors ?? null } : {}),
  };
}

function mapEntrypoint(raw: unknown): ApiEntrypoint {
  const e = raw as Record<string, unknown>;
  return {
    type: asString(e.type, ''),
    configuration: parseConfig(e.configuration),
    dlq: mapDlq(e.dlq),
  };
}

function mapDlq(raw: unknown): { endpoint: string } | null {
  if (!raw || typeof raw !== 'object') return null;
  const d = raw as Record<string, unknown>;
  const endpoint = d.endpoint;
  if (typeof endpoint !== 'string') return null;
  return { endpoint };
}

function mapEndpointGroup(raw: unknown): ApiEndpointGroup {
  const g = raw as Record<string, unknown>;
  return {
    name: asString(g.name, ''),
    type: asString(g.type, 'http-proxy'),
    loadBalancer: mapLoadBalancer(g.loadBalancer),
    endpoints: asArray(g.endpoints).map(mapEndpoint),
    sharedConfiguration: parseConfig(g.sharedConfiguration),
    services: asRecord(g.services),
  };
}

function mapLoadBalancer(raw: unknown): { type: string } {
  if (!raw || typeof raw !== 'object') return { type: 'ROUND_ROBIN' };
  const lb = raw as Record<string, unknown>;
  return { type: asString(lb.type, 'ROUND_ROBIN') };
}

function mapEndpoint(raw: unknown): ApiEndpoint {
  const e = raw as Record<string, unknown>;
  return {
    name: asString(e.name, ''),
    type: asString(e.type, 'http-proxy'),
    weight: asNumber(e.weight, 1),
    inheritConfiguration: asBoolean(e.inheritConfiguration, true),
    secondary: asBoolean(e.secondary, false),
    configuration: parseConfig(e.configuration),
    services: asRecord(e.services),
  };
}

function mapFlow(raw: unknown): ApiFlow {
  const f = raw as Record<string, unknown>;
  return {
    name: asString(f.name, ''),
    enabled: asBoolean(f.enabled, true),
    selectors: asArray(f.selectors),
    request: asArray(f.request).map(mapStep),
    response: asArray(f.response).map(mapStep),
    subscribe: asArray(f.subscribe).map(mapStep),
    publish: asArray(f.publish).map(mapStep),
    tags: asStringArray(f.tags),
  };
}

function mapStep(raw: unknown): ApiFlowStep {
  const s = raw as Record<string, unknown>;
  return {
    name: asString(s.name, ''),
    policy: asString(s.policy, ''),
    enabled: asBoolean(s.enabled, true),
    description: asOptionalString(s.description),
    condition: asOptionalString(s.condition),
    configuration: parseConfig(s.configuration),
  };
}

function mapProperty(raw: unknown): ApiProperty {
  const p = raw as Record<string, unknown>;
  return {
    key: asString(p.key, ''),
    value: asString(p.value, ''),
  };
}

// ── Type-safe helpers ──

function asString(v: unknown, fallback: string): string {
  return typeof v === 'string' ? v : fallback;
}

function asOptionalString(v: unknown): string | undefined {
  return typeof v === 'string' ? v : undefined;
}

function asBoolean(v: unknown, fallback: boolean): boolean {
  return typeof v === 'boolean' ? v : fallback;
}

function asNumber(v: unknown, fallback: number): number {
  return typeof v === 'number' ? v : fallback;
}

function asStringArray(v: unknown): string[] {
  return Array.isArray(v) ? v.filter((item): item is string => typeof item === 'string') : [];
}

function asArray(v: unknown): unknown[] {
  return Array.isArray(v) ? v : [];
}

function asRecord(v: unknown): Record<string, unknown> {
  return typeof v === 'object' && v !== null ? (v as Record<string, unknown>) : {};
}

function parseConfig(config: unknown): Record<string, unknown> {
  if (!config) return {};
  if (typeof config === 'string') {
    try {
      const parsed: unknown = JSON.parse(config);
      return typeof parsed === 'object' && parsed !== null ? (parsed as Record<string, unknown>) : {};
    } catch {
      return {};
    }
  }
  return typeof config === 'object' && config !== null ? (config as Record<string, unknown>) : {};
}
